ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Tables.hs

```diff
@@ -0,0 +1,551 @@
+  , ShowLedgerState (..)
+    -- * UTxO HD
+    -- ** Isolating the tables
+  , InMemory (..)
+  , StowableLedgerTables (..)
+  , TableStuff (..)
+  , TickedTableStuff (..)
+  , mapOverLedgerTables
+  , mapOverLedgerTablesTicked
+  , overLedgerTables
+  , overLedgerTablesTicked
+  , zipOverLedgerTables
+  , zipOverLedgerTablesTicked
+    -- ** Tables values
+  , ApplyMapKind
+  , ApplyMapKind' (..)
+  , MapKind
+  , showsApplyMapKind
+    -- *** Mediators
+  , CodecMK (..)
+  , DiffMK
+  , EmptyMK
+  , IsMapKind (..)
+  , KeysMK
+  , NameMK (..)
+  , QueryMK
+  , SeqDiffMK
+  , TrackingMK
+  , ValuesMK
+    -- ** Serialization
+  , SufficientSerializationForAnyBackingStore (..)
+  , valuesMKDecoder
+  , valuesMKEncoder
+  ) where
+
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Control.Exception as Exn
+import           Data.Kind (Type)
+import qualified Data.Map as Map
+import           Data.Monoid (Sum (..))
+import           NoThunks.Class (NoThunks (..))
+import qualified NoThunks.Class as NoThunks
+
+import           Ouroboros.Consensus.Ticked
+
+import           Ouroboros.Consensus.Storage.LedgerDB.HD.DiffSeq
+
+{-------------------------------------------------------------------------------
+  Basic LedgerState classes
+-------------------------------------------------------------------------------}
+
+class ShowLedgerState (l :: LedgerStateKind) where
+  showsLedgerState :: l (ApplyMapKind' mk) -> ShowS
+
+class ( ShowLedgerState (LedgerTables l)
+      , Eq (l EmptyMK)
+      , Eq (LedgerTables l DiffMK)
+      , Eq (LedgerTables l ValuesMK)
+      ) => TableStuff (l :: LedgerStateKind) where
+
+  data family LedgerTables l :: LedgerStateKind
+
+  -- | Some values of @l mk@ do not determine @mk@, hence the 'SingI' constraint.
+  --
+  -- If it were always the case that @l mk@ not determing @mk@ implies
+  -- @LedgerTables l mk@ also does not determine @mk@, then we would not need
+  -- the 'SingI' constraint. Unfortunately, that is not always the case. The
+  -- example we have found in our prototype UTxO HD implementat is that a Byron
+  -- ledger state does not determine @mk@, but the Cardano ledger tables do.
+  projectLedgerTables :: IsMapKind mk => l mk -> LedgerTables l mk
+
+  -- | Overwrite the tables in some ledger state.
+  --
+  -- The contents of the tables should not be /younger/ than the content of the
+  -- ledger state. In particular, for a
+  -- 'Ouroboros.Consensus.HardFork.Combinator.Basics.HardForkBlock' ledger, the
+  -- tables argument should not contain any data from eras that succeed the
+  -- current era of the ledger state argument.
+  --
+  -- TODO: reconsider the name: don't we use 'withX' in the context of bracket like functions?
+  --
+  -- TODO: This 'IsMapKind' constraint is necessary because the
+  -- 'CardanoBlock' instance uses 'projectLedgerTables'.
+  withLedgerTables :: IsMapKind mk => l any -> LedgerTables l mk -> l mk
```

Comment:
Now that I see the `IsMapKind` constraint here, two things come to mind:
* Do any of the definitions of `projectLedgerTables` or `withLedgerTables` in `TableStuff` instances make use of the functionality that `IsMapKind` provides? If not, we could probably remove them.
* The comments that discuss `IsApplyMapKind` can probably also be removed, since they are outdated

Reply:
We could create a separate issue to find out to investigate my suggestion

Reply:
Can't remove these but I updated the comments.